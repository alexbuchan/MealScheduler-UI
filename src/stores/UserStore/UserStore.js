import Store from '../Store/Store';
import Constants from '../../constants/userConstants';

let userState = {
  auth: false,
  closeFlashMessage: true
};

class UserStore extends Store {
  constructor() {
    super();
  }

  _registerToActions(action) {
    switch(action.actionType) {
      case Constants.REGISTER_USER:
        this.registerUser(action.data);
        break;

      case Constants.LOGIN_USER:
        this.loginUser(action.data);
        break;
      
      case Constants.LOGOUT_USER:
        this.logoutUser(action.data);
        break;

      case Constants.RETRIEVE_USER_DATA_ON_REFRESH:
        this.populateUserDataOnRefresh(action.data);
        break;
    }
  }

  registerUser = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  loginUser = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  populateUserDataOnRefresh = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  logoutUser = (data) => {
    userState.user = null;
    userState.auth = false;
    userState.message = null;
    this.emitChange();
  }

  populateUserData = (data) => {
    userState.auth = data.auth;
    userState.user = data.user;
    userState.message = data.message;
  }

  getUserState() {
    return userState;
  }
}

export default new UserStore();
