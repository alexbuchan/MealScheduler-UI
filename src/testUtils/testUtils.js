import React from 'react';
import { shallow } from 'enzyme';

export const componentSetup = (Component, props = {}, options = {}) => {
  const instance = shallow(<Component { ...props } />, options);
  return instance;
}

export const findByTestAttribute = (instance, attribute) => {
  const wrapper = instance.find(`[data-test="${attribute}"]`);
  return wrapper;
}