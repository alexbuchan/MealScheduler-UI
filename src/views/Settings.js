import React from 'react';
import Form from '../components/Form/Form';
import Input from '../components/Input/Input';
import Checkbox from '../components/Checkbox/Checkbox';
// import Actions from '../actions/Settings/Settings';

class Settings extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: '',
      lightMode: false
    }

    this.submitted = false;
  }

  handleOnChange = (ev) => {
    ev.preventDefault();

    this.setState({
      [ev.target.name]: ev.target.value
    });
  }

  handleLightModeChange = () => {
    this.setState({
      lightMode: !this.state.lightMode
    });
  }

  handleOnSubmit = (fields) => {
    console.log('updateUserSettings() !!!');
    // Actions.updateUserSettings(fields);
  }

  render() {
    return (
      <>
        <h1>Settings</h1>
        <Form
          fields={ this.state }
          onSubmit={ this.handleOnSubmit }
        >
          <Input
            label='Username'
            name='username'
            type='text'
            value={ this.state.username }
            onChange={ this.handleOnChange }
          />

          <Checkbox 
            label='Dark/Light Mode'
            name='lightMode'
            type='checkbox'
            value={ this.state.lightMode }
            onChange={ this.handleLightModeChange }
          />
        </Form>
      </>
    );
  }
}

export default Settings;