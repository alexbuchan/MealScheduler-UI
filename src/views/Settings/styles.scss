.settings-view {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: $contentHeight;

  .settings-header {
    margin-top: 20px;
    margin-bottom: 20px;
  }

  .settings {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100%;
    width: 100%;
  }

  .user-details {
    display: flex;
    justify-content: center;
    align-items: center;

    .user-details-title {
      margin-right: 10px;
    }

    .edit-form-button {
      width: 45px;
      height: 90%;
      color: blue;
      border: 1px solid cornflowerblue;
      border-radius: 4px;
    }

    .cancel-edit-form-button {
      width: 45px;
      height: 90%;
      color: crimson;
      border: 1px solid crimson;
      border-radius: 4px;
    }
  }
}