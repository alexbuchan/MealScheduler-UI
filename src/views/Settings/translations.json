{
  "en": {
    "settings": {
      "page_title": "Settings"
    }
  },

  "es": {
    "settings": {
      "page_title": "Ajustes"
    }
  },

  "fr": {
    "settings": {
      "page_title": "Paramètres"
    }
  },

  "cat": {
    "settings": {
      "page_title": "Configuració"
    }
  }
}