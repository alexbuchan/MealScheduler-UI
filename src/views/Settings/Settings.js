import React from 'react';

// ACTIONS
import Actions from '../../actions/settings/SettingsActions';
import AppActions from '../../actions/app/AppActions';

// STORES
import UserStore from '../../stores/UserStore/UserStore';

// COMPONENTS
import Form from '../../components/Form/Form';
import TextField from '../../components/formComponents/TextField/TextField';
// import FlashMessage from '../../components/FlashMessage/FlashMessage';
import { Dropdown } from 'semantic-ui-react';

// TRANSLATIONS
import translations from './translations.json';
import { translate } from '../../lib/i18n/i18n';
let t = translate(translations);

// Settings View Component
class Settings extends React.Component {
  constructor(props) {
    super(props);

    this.initialState = {
      username: UserStore.getUser().username,
      email: UserStore.getUser().email,
      edit: false,
      disableSubmitButton: true
    }

    this.state = this.initialState;
    this.submitted = false;
    this.t = t(props.props.locale);
  }

  _onChange = () => {
    if (UserStore.getUser()) {
      this.setState({
        username: UserStore.getUser().username,
        email: UserStore.getUser().email
      });

      this.initialState = this.state;
    }
  }

  componentDidMount() {
    this.t = t(this.props.props.locale);
    UserStore.addChangeListener(this._onChange);
  }

  componentWillUnmount() {
    UserStore.removeChangeListener(this._onChange);
  }

  handleEditClick = (ev) => {
    this.setState({
      edit: true,
      disableSubmitButton: false
    });
  }

  renderEditClick = () => {
    if (!this.state.edit) {
      return <button data-test='edit-button' className="edit-form-button" onClick={ this.handleEditClick }>Edit</button>;
    }

    return null;
  }

  handleCancelEditClick = (ev) => {
    this.setState(this.initialState);
  }

  renderCancelEditClick = () => {
    if (this.state.edit) {
      return <button data-test='cancel-edit-button' className="cancel-edit-form-button" onClick={ this.handleCancelEditClick }>X</button>;
    }

    return null;
  }

  handleOnChange = (ev) => {
    ev.preventDefault();

    this.setState({
      [ev.target.name]: ev.target.value
    });
  }

  handleOnSubmit = () => {
    const fields = { username: this.state.username, email: this.state.email };
    Actions.updateUserSettings(fields);
    this.setState(this.initialState);
  }

  handleLocaleChange = (_, { value }) => {
    AppActions.changeLocale(value);
  }

  appLocales = () => {
    return this.props.props.locales.map(locale => { return { text: locale, value: locale } });
  }

  render() {
    const fields = { username: this.state.username, email: this.state.email };

    return (
      <div className="settings-view">
        <h1 className="settings-header">{ this.t('settings.page_title') }</h1>

        <div className="settings">
          <div className="user-signup">
            <div className="user-details">
              <h4 className="user-details-title">User Details</h4>
              { this.renderEditClick() }
              { this.renderCancelEditClick() }

            </div>

            <Form
              disableSubmitButton={ this.state.disableSubmitButton }
              validate={ true }
              fields={ fields }
              areRequired={ ['username', 'email'] }
              onSubmit={ this.handleOnSubmit }
            >
              <TextField
                label='Username'
                name='username'
                type='text'
                value={ this.state.username }
                disabled={ !this.state.edit }
                onChange={ this.handleOnChange }
                isRequired={ true }
              />

              <TextField
                label='Email'
                name='email'
                type='text'
                value={ this.state.email }
                onChange={ this.handleOnChange }
                disabled={ !this.state.edit }
                isRequired={ true }
              />
            </Form>
          </div>

          <Dropdown
            placeholder='Language'
            fluid
            search
            selection
            value={ this.props.props.locale }
            options={ this.appLocales() }
            onChange={ this.handleLocaleChange }
          />
        </div>
      </div>
    );
  }
}

export default Settings;