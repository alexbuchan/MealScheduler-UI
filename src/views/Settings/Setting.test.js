import Settings from './Settings';
import UserStore from '../../stores/UserStore/UserStore';
import { componentSetup } from '../../testUtils/testUtils';

describe('Settings', () => {
  beforeEach(() => {
    const data = {
      auth: true,
      message: 'User logged in',
      user: {
        username: 'user',
        email: 'user@email.com'
      }
    };

    UserStore.loginUser(data);
  });

  describe('#render', () => {

  });

  describe('#componentWillUnmount', () => {
    let instance;
  
    beforeEach(() => {
      instance = componentSetup(Settings);
    });
    
    it('Should remove the Store event listener', () => {
      const removeChangeListenerSpy = jest.spyOn(UserStore, 'removeChangeListener');
      instance.unmount();
      expect(removeChangeListenerSpy).toHaveBeenCalled();
    });
  });
});