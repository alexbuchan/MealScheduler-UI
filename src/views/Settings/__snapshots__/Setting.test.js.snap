// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings #render renders correctly 1`] = `
<div
  className="settings-view"
>
  <h1
    className="settings-header"
  >
    Settings
  </h1>
  <div
    className="settings"
  >
    <div
      className="user-signup"
    >
      <div
        className="user-details"
      >
        <h4
          className="user-details-title"
        >
          User Details
        </h4>
        <button
          className="edit-form-button"
          data-test="edit-button"
          onClick={[Function]}
        >
          Edit
        </button>
      </div>
      <Form
        areRequired={
          Array [
            "username",
            "email",
          ]
        }
        disableSubmitButton={true}
        fields={
          Object {
            "email": "user@email.com",
            "username": "user",
          }
        }
        onSubmit={[Function]}
        redirect={false}
        submitButtonLabel="Submit"
        validate={true}
      >
        <TextField
          disabled={true}
          isRequired={true}
          label="Username"
          name="username"
          onChange={[Function]}
          type="text"
          value="user"
        />
        <TextField
          disabled={true}
          isRequired={true}
          label="Email"
          name="email"
          onChange={[Function]}
          type="text"
          value="user@email.com"
        />
      </Form>
    </div>
  </div>
</div>
`;
